Enable `cirq.num_qubits` protocol with `Gate` classes

BillGatesNephew opened the pull request:
Enables usage of the `cirq.num_qubits` protocol with `Gate` classes per requested in #3341.

mpharrigan commented:
This PR would be extremely compelling if you add a test that would have failed without the fix!

mpharrigan commented:
In the linked issue, we were discussing using @staticmethod decorators. Is there a reason you didn't use that route?

BillGatesNephew commented:
@mpharrigan I added a test that checks a few different `Gate` classes with the updated functionality. Also, I initially thought about going that route, but I think it would've required changing the declaration of every `_num_qubits_`  from `_num_qubits_(self)` to `_num_qubits_(cls)`. I also wasn't sure how well the addition of `@staticmethod` would play with the decorators on the `_num_qubits_` function in `Gate`: 
```python
@value.alternative(requires='num_qubits', implementation=_num_qubits_proto_from_num_qubits)
@value.alternative(requires='_qid_shape_', implementation=_num_qubits_from_shape)
```
because both of the implementation parameters are instance methods. Because of this, I figured the route mentioned in the original issue would've ended up touching a lot of files. I just figured that it would be better to contain the solution to one function than make changes to every file where `_num_qubits_` was defined.

mpharrigan reviewed (changes requested):
Only certain classes know their number of qubits statically. For example: this code breaks

```python
cirq.num_qubits(cirq.MeasurementGate)
```

BillGatesNephew commented:
@mpharrigan I added a potential change to try calling the function if it is defined statically, and catch a raised exception to return `NotImplemented`. I also added additional assertions to check this case raises the appropriate exception. Generally, I try to avoid `try: ... except: ...` blocks, but to keep this change contained to a single file I added one. I'm not sure how you guys feel about them, and if they're frowned upon, then I can look into the potential `@staticmethod`solution.

mpharrigan commented:
There's nothing wrong with try/except blocks; but there's also nothing wrong with a change touching multiple files :) 

I think the staticmethod decorators would be more idiomatic. As a side benefit, you could call the method directly e.g. `cirq.XPowGate._num_qubits_()`

viathor commented on cirq/protocols/qid_shape_protocol.py:
```diff
@@ -160,14 +162,27 @@ def num_qubits(
+
+        try:
+            return opt_getter_func(None)
+        except AttributeError:
```
I'm a little leery about using a try-except block here since `AttributeError` could be raised for reasons other than the one we have in mind here. It'd be safer to directly check the presence of the attribute using `hasattr`.

(This is similar to the EAFP idiom for checking key presence in a dict, so it's certainly idiomatic. It seems unnecessarily risky, though.)

mpharrigan commented:
@BillGatesNephew are you still interested in working on this PR?

mpharrigan commented:
closing as stale, feel free to re-open
